sponsorship_compassion/model/contracts.py

```diff
@@ -61,8 +67,55 @@ def _get_contract_from_invoice(invoice_obj, cr, uid, invoice_ids,
                         contract.write({
                             'activation_date': datetime.today().strftime(DF)})
 
+                        # Cancel the old invoices if a contract is activated
+                        self._cancel_old_invoices(
+                            cr, uid,
+                            invoice.partner_id.id,
+                            contract.id,
+                            first_pay_date,
+                            context)
         return list(res)
 
+    def _cancel_old_invoices(
+            self, cr, uid, partner_id,
+            contract_id, date_invoice, context=None):
+        invoice_line_obj = self.pool.get('account.invoice.line')
+        invoice_obj = self.pool.get('account.invoice')
+        invoice_ids = invoice_obj.search(
+            cr, uid,
+            [('partner_id', '=', partner_id),
+             ('state', '=', 'open'),
+             ('date_invoice', '<', date_invoice)
+             ],
+            context=context)
+
+        for invoice in invoice_obj.browse(cr, uid, invoice_ids, context):
+            invoice_lines = invoice.invoice_line
+            contract_ids = [
+                invl.contract_id.id for invl in invoice_lines
+                if invl.contract_id]
+            contract_ids = list(set(contract_ids))
+
+            wf_service = netsvc.LocalService('workflow')
+            if contract_ids and contract_id in contract_ids:
+                if len(contract_ids) == 1:
+                    wf_service.trg_validate(uid, 'account.invoice',
+                                            invoice.id, 'invoice_cancel', cr)
+                else:
+                    invoice_obj.action_cancel_draft(
+                        cr, uid, invoice.id, context)
+
+                    inv_line_ids = invoice_line_obj.search(
+                        cr, uid,
+                        [('contract_id', '=', contract_id)])

```

Comment:
This will unlink all invoice lines attached to this contract ! You should limit to those in the current invoice.